courses/utils.py

```diff
@@ -1,251 +1,87 @@
-from courses.models import CourseRun, CourseRunGrade, CourseRunCertificate
-from users.api import User, fetch_user
+from courses.models import CourseRunGrade, CourseRunCertificate
 
-logger = logging.getLogger(__name__)
 
-
-def _verify_configuration(override_grade=None, user=None):
-    """
-    Verify if required variables are setup properly and the command arguments are correct
-    """
-    if settings.OPENEDX_GRADES_API_TOKEN is None:
-        raise ImproperlyConfigured("OPENEDX_GRADES_API_TOKEN is not set")
-
-    if settings.OPENEDX_API_KEY is None:
-        raise ImproperlyConfigured("OPENEDX_API_KEY is not set")
-
-    # Grade override for all users for the course run. Not allowed.
-    if override_grade and not user:
-        exception = PermissionDenied(
-            "No user supplied with override grade. Force overwrite of grade is not supported \
-                for all users. Grade should only be supplied when a specific user is targeted."
-        )
-        logger.error(str(exception))
-        raise exception
-
-
-def _get_course_run(readable_id):
+def ensure_course_run_grade(user, course_run, edx_grade, should_update=False):
     """
-    Get the course run needed
-    """
-    course_run = CourseRun.objects.get(courseware_id=readable_id)
-
-    if not course_run:
-        exception = CourseRun.DoesNotExist(
-            "Could not find course run with courseware_id={}".format(readable_id)
-        )
-        logger.error(str(exception))
-        raise exception
-    return course_run
-
-
-def _generate_grade(user, course_run, edx_grade, force=False):
-    """
-    Create a local CourseRunGrade record with the passed values.
+    Ensure that the local grades repository has the grade for the User/CourseRun combination supplied.
 
     Args:
-        user (User): The user for which the grade should be created
-        course_run (CourseRun): The course run which the grade is being created
-        edx_grade (edx_api.grades.CurrentGrade): The grade object retrieved from OpenEdX
-        force (bool): Overwrite the local CourseRunGrade record if it exists
+        user (user.models.User): The user for whom the grade is being synced
+        course_run (courses.models.CourseRun): The course run for which the grade is created
+        edx_grade (edx_api.grades.models.UserCurrentGrade): The OpenEdx grade object
+        should_update (bool): Update the local grade record if it exists
+
+    Returns:
+        (courses.models.CourseRunGrade, bool, bool) that depicts the CourseRunGrade, created and updated values
     """
-    if (
-        CourseRunGrade.objects.filter(user=user, course_run=course_run).exists()
-        and not force
-    ):
-        logger.warning(
-            "Grade exists for user: %s and courseware_id: %s. Not overwriting",
-            user.email,
-            course_run.courseware_id,
-        )
+    grade_qset = CourseRunGrade.objects.filter(course_run=course_run, user=user)
```

Comment:
`get` instead of `filter`

Reply:
But would be good to ensure a single record against filter returned.

Reply:
I just rewrote the method anyway to utlize `get_or_create` and `update_or_create` to simplify it without the need for multi-step checks.